uuid-types/src/Data/UUID/Types/Internal.hs

```diff
@@ -374,60 +378,34 @@ toString uuid = hexw0 w0 $ hexw1 w1 ""
 fromText :: Text -> Maybe UUID
 fromText = fromASCIIBytes . T.encodeUtf8
 
--- | Convert a UUID into a hyphentated string using lower-case letters.
+-- | Convert a UUID into a hyphenated string using lower-case letters.
 toText :: UUID -> Text
 toText = T.decodeLatin1 . toASCIIBytes
 
--- | Convert a UUID into a hyphentated string using lower-case letters, packed
+-- | Convert a UUID into a hyphenated string using lower-case letters, packed
 --   as ASCII bytes into `B.ByteString`.
 --
 --   This should be equivalent to `toString` with `Data.ByteString.Char8.pack`.
 toASCIIBytes :: UUID -> B.ByteString
-toASCIIBytes uuid = BI.unsafeCreate 36 (pokeASCII uuid)
-
--- | Helper function for `toASCIIBytes`
-pokeASCII :: UUID -> Ptr Word8 -> IO ()
-pokeASCII uuid ptr = do
-    pokeDash 8
-    pokeDash 13
-    pokeDash 18
-    pokeDash 23
-    pokeSingle 0  w0
-    pokeDouble 9  w1
-    pokeDouble 19 w2
-    pokeSingle 28 w3
+toASCIIBytes uuid = BBI.unsafeCreate 36 (BBPI.runF hyphenatedUUIDFixedPrim uuid)
+
+hyphenatedUUIDFixedPrim :: BBP.FixedPrim UUID
+hyphenatedUUIDFixedPrim = uuidToByteTuples BBP.>$< wordFixedPrim
   where
-    (w0, w1, w2, w3) = toWords uuid
+    wordFixedPrim :: BBP.FixedPrim (Word32, (Word16, (Word16, (Word16, (Word16, Word32)))))
+    wordFixedPrim = BBP.word32HexFixed BBP.>*<
```

Comment:
Note to self,  the `word<N>HexFixed` calls into C function in `bytestring`:

```c
char* _hs_bytestring_uint_hex (unsigned int x, char* buf) {
    // write hex representation in reverse order
    char c, *ptr = buf, *next_free;
    do {
        *ptr++ = digits[x & 0xf];
        x >>= 4;
    } while ( x );
    // invert written digits
    next_free = ptr--;
    while(buf < ptr) {
        c      = *ptr;
        *ptr-- = *buf;
        *buf++ = c;
    }
    return next_free;
};
```

Reply:
No, that's wrong, the `Fixed` variants just split the number into halves:

https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/src/Data.ByteString.Builder.Prim.ASCII.html#word32HexFixed

```
-- | Encode a 'Word8' using 2 nibbles (hexadecimal digits).
{-# INLINE word8HexFixed #-}
word8HexFixed :: FixedPrim Word8
word8HexFixed = fixedPrim 2 $ \x op -> do
  enc <- encode8_as_16h lowerTable x
  unalignedWriteU16 enc op

-- | Encode a 'Word16' using 4 nibbles.
{-# INLINE word16HexFixed #-}
word16HexFixed :: FixedPrim Word16
word16HexFixed =
    (\x -> (fromIntegral $ x `shiftR` 8, fromIntegral x))
      >$< pairF word8HexFixed word8HexFixed

-- | Encode a 'Word32' using 8 nibbles.
{-# INLINE word32HexFixed #-}
word32HexFixed :: FixedPrim Word32
word32HexFixed =
    (\x -> (fromIntegral $ x `shiftR` 16, fromIntegral x))
      >$< pairF word16HexFixed word16HexFixed

-- | Encode a 'Word64' using 16 nibbles.
{-# INLINE word64HexFixed #-}
word64HexFixed :: FixedPrim Word64
word64HexFixed =
    (\x -> (fromIntegral $ x `shiftR` 32, fromIntegral x))
      >$< pairF word32HexFixed word32HexFixed
```

Reply:
So that's what https://github.com/haskell-hvr/uuid/pull/80#issuecomment-2005708058 was saying